lib/dao.js

```diff
@@ -2421,6 +2421,108 @@ DataAccessObject.updateAll = function(where, data, options, cb) {
   return cb.promise;
 };
 
+DataAccessObject.patchById = function patchById(id, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+  assert(arguments.length >= 3, 'At least three arguments are required');
+
+  var Model = this;
+  var where = {};
+  var idPropertyName = idName(Model);
+
+  where[idPropertyName] = id;
+  if (cb === undefined) {
+    if (typeof options === 'function') {
+      // patchById(id, data, cb)
+      cb = options;
+      options = {};
+    }
+  }
+
+  data = data || {};
+  cb = cb || utils.createPromiseCallback();
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+    'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = Model.getDataSource().connector;
+  assert(typeof connector.update === 'function',
+    'update() must be implemented by the connector');
+
+  var hookState = {};
+  this.applyProperties(data);
+
+  var context = {
+    Model: Model,
+    instanceId: id,
+    data: data,
+    hookState: hookState,
+    options: options,
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
+    doUpdate(ctx.instanceId, ctx.data);
+  });
+  function doUpdate(instanceId, data) {
+    try {
+      data = removeUndefined(data);
+      data = Model._coerce(data);
+    } catch (err) {
+      return process.nextTick(function() {
+        cb(err);
+      });
+    }
+
+    function patchCallback(err, info) {
+      if (err) return cb(err);
+      var ctx = {
+        Model: Model,
+        data: data,
+        hookState: hookState,
+        options: options,
+      };
+      Model.notifyObserversOf('loaded', ctx, function(err) {
+        if (err) return cb(err);
+        var context = {
+          Model: Model,
+          isNewInstance: false,
+          hookState: hookState,
+          options: options,
```

Comment:
I see.

@raymondfeng what's your opinion here? In 3.0, I was thinking about remapping `PATCH /api/my-model/:id` from `prototype.updateAttributes` to `patchById` (`PUT /api/my-model/:id` will be mapped to `replaceById`), so that we can avoid the extra database round-trip which fetches the instance before it's partially updated. However, assuming the current connector implementation, we won't be able to return the updated instance in HTTP response.

To give more context, we have already introduced `replaceById` which should be used in most scenarios where `prototype.updateAttributes` was used now and `replaceById` returns full instance data. As I see it, `patchById` should be used only in specific cases where users intentionally want to do a partial update (or perhaps use MongoDB extended operators), in which case it may be ok to not get back the fully updated instance. I don't think `updateAttributes` is able to parse & apply all MongoDB extended operations, which means the return value is inaccurate anyways. And also if NoSQL databases don't provide an atomic way how to make a partial update and get the full updated object back, then I think it's better to model our API with this limitation in mind and don't pretend we can get accurate updated data when that's not always true.

Having wrote that, I think it's ok to let `PATCH /api/mymodel/:id` to return only an `info` object like `{ count: 1 }`.

**LB 2.x**

`PUT /api/mymodel/:id` calls `prototype.updateAttributes` (partial update) and returns full instance data

**LB 3.x as proposed**

`PUT /api/mymodel/:id` calls `replaceById` (full update/replace) and returns full instance data
`PATCH /api/mymodel/:id` calls `patchById` (partial update) and returns `info` only

Alternatively, we could extend connectors to provide API where we can pass `id` and partial `data` and get back full data of the updated instance (this can be either a flag for `updateAttributes` or a new method), but I think that's a lot of extra work and not all databases provide atomic API that we could leverage.

Thoughts?

I am not sure if my comment makes sense. If it does not, then let's discuss next week on Slack.